### PR TITLE
Review of endpoint to return all resources

### LINKED ISSUE #379: Review of endpoint to return all resources
Review of the _**/extjs/search/list**_ endpoint, in particular the filters support and properties returned in response payload:

- [x] Add missing order/sort parameter with ascending/descending option and make it supported for all available resource attributes such as: name, description, owner, modification/creation date
- [x] Add missing support for filtering resources by: 
     - [x] _creator_ (the owner of the resource)
     - [x] _editor_ (the last user who modified the resource that is used for indicating the _modified by_ on the front-end) 
- [x] Add missing support for filtering resources by group (independently if a group has W or R permission). For example, giving one or more group names, GeoStore returns all resources having those groups in their security rules
- [x] The resource does not contain permissions information: _canCopy_, _canDelete_, _canEdit_. This information should be included in the response
- [x] Filtering by _Creation date_ and _Last update_ is not working correctly with filters GREATER_THAN_OR_EQUAL_TO and LESS_THAN_OR_EQUAL_TO e.g.:

```
<AND>
  <FIELD>
    <field>CREATION</field>
    <operator>GREATER_THAN_OR_EQUAL_TO</operator>
    <value>2024-06-01T00:00:00</value>
  </FIELD>
  <FIELD>
    <field>CREATION</field>
    <operator>LESS_THAN_OR_EQUAL_TO</operator>
    <value>2024-07-27T23:59:59</value>
  </FIELD>
</AND>
```

**Note:** for all the above points it is not requested to change the filtering strategies currently adopted.

### PR DESCRIPTION
**Pull request by axl8713**
Closes #379 .

### CODE REVIEW

File: src/core/model/src/main/java/it/geosolutions/geostore/core/model/Resource.java [hunk 1]
```diff
@@ -119,6 +120,10 @@ public class Resource implements Serializable, CycleRecoverable {
     @OneToMany(mappedBy = "resource", cascade = CascadeType.REMOVE, fetch = FetchType.LAZY)
     private List<SecurityRule> security;
 
+    @Transient private boolean canEdit;
+    @Transient private boolean canDelete;
+    @Transient private boolean canCopy;
```
Comment (afabiani) [inline src/core/model/src/main/java/it/geosolutions/geostore/core/model/Resource.java:125]: You’re mixing persistence fields with UI/permission flags in the same entity.
Consider documenting clearly that these fields are only set at runtime, or wrap the Resource in a “ResourceView” object so that the entity remains focused on business/persistence data, and UI permission logic lives elsewhere.

File: src/core/services-api/src/main/java/it/geosolutions/geostore/services/ResourceService.java [hunk 2]
```diff
@@ -200,6 +200,29 @@ List<Resource> getResources(
...
+    List<Resource> getResources(
+            SearchFilter filter,
+            Integer page,
+            Integer entries,
+            String sortBy,
+            String sortOrder,
+            boolean includeAttributes,
+            boolean includeData,
+            User authUser)
+            throws BadRequestServiceEx, InternalErrorServiceEx;
```
Comment (afabiani) [inline src/core/services-api/src/main/java/it/geosolutions/geostore/services/ResourceService.java:224]: * If you foresee further enhancements (e.g., multiple sort fields), consider a SearchRequest or ResourceQuery object to encapsulate all pagination + sorting + filtering parameters. It keeps the method signature clean and reduces the risk of “parameter bloat.”

* Document or validate that sortBy must match a valid entity field. Right now, you only do a sortBy != null && !sortBy.isBlank(), but if the user passes an invalid field, it may throw an exception deeper in the DAO layer.

File: src/core/services-impl/src/main/java/it/geosolutions/geostore/services/ResourceServiceImpl.java [hunk 3]
```diff
@@ -491,72 +493,82 @@ public long getCountByFilter(SearchFilter filter)
...
-                            }
-                        }
-                    }
-                    if (!authUserIsOwner) resourceCanBeListed = resource.isAdvertised();
+    private List<ShortResource> convertToShortList(List<Resource> resources, User user) {
+        return resources.stream()
+                .filter(r -> isResourceCanBeListed(user, r))
+                .map(r -> createShortResourceWithPermissions(user, r))
+                .collect(Collectors.toList());
+    }
```
Comment (afabiani) [inline src/core/services-impl/src/main/java/it/geosolutions/geostore/services/ResourceServiceImpl.java:505]: Overall, streaming is fine here. Just watch out for repeated database calls in .filter(...) or .map(...). If isResourceCanBeListed(...) or createShortResourceWithPermissions(...) each cause multiple DB queries, you might degrade performance.

Confirm that you’re not repeatedly calling resourceService.getSecurityRules(r.getId()) inside this stream for many resources. You might want to batch fetch or join if performance becomes an issue.

File: src/core/services-impl/src/main/java/it/geosolutions/geostore/services/ResourceServiceImpl.java [hunk 4]
```diff
@@ -491,72 +493,82 @@ public long getCountByFilter(SearchFilter filter)
...
 
-    public static List<String> extractGroupNames(Set<UserGroup> groups) {
-        List<String> groupNames = new ArrayList<String>();
-        if (groups == null) {
-            return groupNames;
+    /**
+     * Checks if the user can modify and delete the loaded resource (and associated attributes and
+     * stored data). This to inform the client in HTTP response result.
+     */
+    public ShortResource addPermissionsToShortResource(
```
Comment (afabiani) [inline src/core/services-impl/src/main/java/it/geosolutions/geostore/services/ResourceServiceImpl.java:541]: Possible Improvement:
Unify permission logic in a single service or utility. E.g., a PermissionService that says hasWritePermission(User, Resource). This ensures you don’t accidentally diverge logic in multiple places.

File: src/modules/rest/extjs/src/main/java/it/geosolutions/geostore/services/rest/impl/RESTExtJsServiceImpl.java [hunk 5]
```diff
@@ -353,29 +385,47 @@ public ExtResourceList getExtResourcesList(
...
+     * @param resource
+     * @param user
+     * @return <code>true</code> if the resource should be visible to the user, <code>false</code>
+     *     otherwise
+     */
+    private boolean isResourceAvailableForUser(Resource resource, User user) {
+        return resource.isAdvertised()
+                || user.getRole().equals(Role.ADMIN)
+                || isUserOwner(user, resource);
+    }
```
Comment (afabiani) [inline src/modules/rest/extjs/src/main/java/it/geosolutions/geostore/services/rest/impl/RESTExtJsServiceImpl.java:414]: Nicely broken down into smaller private methods, good use of naming.
Potential Drawback: You still call resourceService.getSecurityRules(r.getId()) inside isUserOwner(...), which can lead to an additional DB call per resource. This might be fine for small sets, but for large sets, consider a more optimized approach or a single query that fetches all relevant rules in one go.

File: src/modules/rest/extjs/src/main/java/it/geosolutions/geostore/services/rest/impl/RESTExtJsServiceImpl.java [hunk 6]
```diff
@@ -280,17 +296,26 @@ public String getResourcesByCategory(
...
+     * java.lang.Integer, java.lang.String, java.lang.String, boolean, boolean, it.geosolutions.geostore.services.dto.search.SearchFilter)
      */
+
     @Override
     public ExtResourceList getExtResourcesList(
             SecurityContext sc,
             Integer start,
             Integer limit,
+            String sortBy,
+            String sortOrder,
```
Comment (afabiani) [inline src/modules/rest/extjs/src/main/java/it/geosolutions/geostore/services/rest/impl/RESTExtJsServiceImpl.java:318]: Similar to the approach mentioned earlier, a single request object could simplify the parameter list. This also future-proofs the code for additional flags in the future.

File: src/modules/rest/extjs/src/main/java/it/geosolutions/geostore/services/rest/impl/RESTExtJsServiceImpl.java [hunk 7]
```diff
@@ -327,9 +352,16 @@ public ExtResourceList getExtResourcesList(
 
         try {
             List<Resource> resources =
-                    filterOutUnadvertisedResources(
+                    handleFoundResources(
```
Comment (afabiani) [inline src/modules/rest/extjs/src/main/java/it/geosolutions/geostore/services/rest/impl/RESTExtJsServiceImpl.java:355]: Essentially the same pattern repeated: check availability + set permission flags. Consider combining or reusing code from ResourceServiceImpl to avoid duplication. Possibly move “permission setting” logic to a single place so that the same user-resource checks are consistent across the application.

File: src/modules/rest/extjs/src/main/java/it/geosolutions/geostore/services/rest/impl/RESTExtJsServiceImpl.java [hunk 8]
```diff
@@ -353,29 +385,47 @@ public ExtResourceList getExtResourcesList(
...
+
+    private boolean isUserOwner(User authUser, Resource r) {
+        return resourceService.getSecurityRules(r.getId()).stream()
+                .map(SecurityRule::getUser)
+                .filter(Objects::nonNull)
+                .anyMatch(ruleUser -> ruleUser.getId().equals(authUser.getId()));
+    }
+
+    private Resource addPermissionsToResource(Resource resource, User user) {
+        ResourceEnvelop envelope = new ResourceEnvelop(resource, user);
```
Comment (afabiani) [inline src/modules/rest/extjs/src/main/java/it/geosolutions/geostore/services/rest/impl/RESTExtJsServiceImpl.java:424]: This internal class encapsulates a chunk of permission logic. Some logic duplicates the code in ResourceServiceImpl and RESTExtJsServiceImpl.
If ResourceEnvelop is primarily for permission decoration, consider a single approach: either do it in the service layer, or do it in a dedicated “decorator” layer. Minimizing overlap will make it easier to maintain.

File: src/modules/rest/extjs/src/main/java/it/geosolutions/geostore/services/rest/impl/RESTExtJsServiceImpl.java [hunk 9]
```diff
@@ -705,60 +756,62 @@ private void readSecurity() {
...
                 }
             }
         }
 
+        private boolean canUserGroupsEditResource() {
+            List<String> groupsNames =
+                    authUser.getGroups().stream()
+                            .map(UserGroup::getGroupName)
+                            .collect(Collectors.toList());
+            return resourceService.getGroupSecurityRule(groupsNames, r.getId()).stream()
```
Comment (afabiani) [inline src/modules/rest/extjs/src/main/java/it/geosolutions/geostore/services/rest/impl/RESTExtJsServiceImpl.java:777]: This is a direct DB call for each resource. If getExtResourcesList() is returning hundreds or thousands of resources, each resource may lead to repeated queries.
Evaluate how frequently you call these “per resource” methods. If the code is triggered in a big listing, you may want a bulk approach to fetch all SecurityRules relevant to that batch of resources.

File: src/core/services-impl/src/main/java/it/geosolutions/geostore/util/SearchConverter.java [hunk 10]
```diff
@@ -200,7 +203,7 @@ public void visit(FieldFilter filter) throws InternalErrorServiceEx {
         f.setOperator(op);
 
         if (type == Date.class) {
-            SimpleDateFormat sdf = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss.SSS");
+            SimpleDateFormat sdf = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss");
```
Comment (afabiani) [inline src/core/services-impl/src/main/java/it/geosolutions/geostore/util/SearchConverter.java:206]: SimpleDateFormat is not thread-safe. Currently, you use it in a single place (which might be okay). But in multi-threaded contexts or if SearchConverter is used as a singleton, you could get concurrency issues.

Use DateTimeFormatter (Java 8+).

File: src/core/services-impl/src/main/java/it/geosolutions/geostore/util/SearchConverter.java [hunk 11]
```diff
@@ -242,6 +241,36 @@ public void visit(CategoryFilter filter) throws InternalErrorServiceEx {
...
+
+        Filter f = new Filter();
+        f.setOperator(op);
+        f.setProperty("security.group.groupName");
+
+        List<String> names = filter.getNames();
+
+        if (SearchOperator.IN != filter.getOperator() && names.size() != 1) {
+            throw new IllegalStateException("Erroneous search op " + filter.getOperator());
+        }
```
Comment (afabiani) [inline src/core/services-impl/src/main/java/it/geosolutions/geostore/util/SearchConverter.java:263]: For IN, you require names.size() != 1. This is good, but you might also want to handle “empty list” vs. “null.”

You do have tests for GroupFilter in RESTExtJsServiceImplTest. Make sure you also test edge cases: empty names, null operator, etc.

File: src/modules/rest/extjs/src/test/java/it/geosolutions/geostore/services/rest/impl/RESTExtJsServiceImplTest.java [hunk 12]
```diff
@@ -362,14 +376,431 @@ public void testGetAllResources_unadvertised() throws Exception {
         }
     }
 
+    @Test
```
Comment (afabiani) [inline src/modules/rest/extjs/src/test/java/it/geosolutions/geostore/services/rest/impl/RESTExtJsServiceImplTest.java:379]: Great coverage for new search/filter/sort features, verifying the new code paths (sorting by date, filtering by group, date-based filters, etc.).

Overall, these tests do a good job hitting the main functionality. Just confirm you also test negative scenarios or edge conditions that might raise exceptions (BadRequestServiceEx, IllegalStateException, etc.).

Suggestions for Additional Test Cases:

- Empty Results: Check what happens if no resource matches the filter.
- Invalid sortBy: If the user tries sorting by a non-existing field.
- Null filter or filter with null names in GroupFilter.
- Security Edge Cases: A resource that has no security rules, or a user with no groups, etc.

**Review by afabiani — COMMENTED** [review]
General Observations
1. Repeated Permission Logic

- You have permission checks scattered in ResourceServiceImpl, RESTExtJsServiceImpl, ResourceEnvelop, and the ShortResource creation. This can lead to maintenance headaches.
- Recommendation: Centralize read/write/copy permission checks in a single place (e.g., PermissionService or a “Policy” class). Then, everything that needs permission info calls that service.

2. Potential Performance / DB Queries

- Repeated calls to resourceService.getSecurityRules(...) inside loops or streams can become expensive if you have many resources.
- Recommendation: Where feasible, fetch all relevant SecurityRules for the user in bulk. Or do a single join in the DAO so each Resource already has the needed rules loaded.

3. DTO vs. Entity

- The @Transient fields (canEdit, canDelete, canCopy) indicate you’re mixing view-layer logic with domain objects. If that’s a conscious design choice, it’s valid, but watch out for “entity bloat.”
- Recommendation: A simpler approach might be returning a dedicated ResourceDTO from the service layer, which includes these transient fields and no JPA annotations.

4. Logging / Exception Handling

- Some methods (like search(Search searchCriteria)) catch IllegalArgumentException and rethrow a BadRequestServiceEx with “Resource search failed.” That’s fine, but ensure you log the root cause somewhere for debugging.
- Recommendation: Add a logger statement with the exception’s message or full stack trace if it helps with debugging.

5. Documentation

- Before consider this PR done, let's make sure we updated the WIKI API sections with the relevant parameters.
- There shouldn't be any model update here, so no need to update the `doc/sql/` scripts

File: src/core/services-impl/src/main/java/it/geosolutions/geostore/services/ResourceServiceImpl.java (hunk 3 above)
Comment (axl8713) [reply src/core/services-impl/src/main/java/it/geosolutions/geostore/services/ResourceServiceImpl.java]: Fixed. Thanks!

File: src/modules/rest/extjs/src/main/java/it/geosolutions/geostore/services/rest/impl/RESTExtJsServiceImpl.java (hunk 5 above)
Comment (axl8713) [reply src/modules/rest/extjs/src/main/java/it/geosolutions/geostore/services/rest/impl/RESTExtJsServiceImpl.java]: Fixed. Thanks!

File: src/modules/rest/extjs/src/main/java/it/geosolutions/geostore/services/rest/impl/RESTExtJsServiceImpl.java (hunk 9 above)
Comment (axl8713) [reply src/modules/rest/extjs/src/main/java/it/geosolutions/geostore/services/rest/impl/RESTExtJsServiceImpl.java]: Fixed. Thanks!

File: src/core/services-impl/src/main/java/it/geosolutions/geostore/util/SearchConverter.java (hunk 10 above)
Comment (axl8713) [reply src/core/services-impl/src/main/java/it/geosolutions/geostore/util/SearchConverter.java]: Changed to use DateTimeFormatter. Thanks!

File: src/core/services-impl/src/main/java/it/geosolutions/geostore/util/SearchConverter.java (hunk 11 above)
Comment (axl8713) [reply src/core/services-impl/src/main/java/it/geosolutions/geostore/util/SearchConverter.java:271]: This condition is actually checking filter values for operators that are not `IN` (e.g. for `EQUAL_TO` assures that we have just one value to avoid ambiguities).
Added some edge cases tests anyway.

File: src/modules/rest/extjs/src/test/java/it/geosolutions/geostore/services/rest/impl/RESTExtJsServiceImplTest.java (hunk 12 above)
Comment (axl8713) [reply src/modules/rest/extjs/src/test/java/it/geosolutions/geostore/services/rest/impl/RESTExtJsServiceImplTest.java:380]: > Empty Results: Check what happens if no resource matches the filter.

Added test cases to check empty result in this case.

> Invalid sortBy: If the user tries sorting by a non-existing field.

The exception thrown here is already handled and silenced by the REST service method, returning `null`. Added test anyway.

> Null filter or filter with null names in GroupFilter.

Added specific tests, see [here](https://github.com/geosolutions-it/geostore/pull/392#discussion_r1920268922).

> Security Edge Cases: A resource that has no security rules, or a user with no groups, etc.

Added specific tests for other corner cases.

File: src/core/services-impl/src/main/java/it/geosolutions/geostore/services/ResourceServiceImpl.java (hunk 4 above)
Comment (axl8713) [reply src/core/services-impl/src/main/java/it/geosolutions/geostore/services/ResourceServiceImpl.java]: Created a specific `PermissionService` to handle resource access checks.

File: src/modules/rest/extjs/src/main/java/it/geosolutions/geostore/services/rest/impl/RESTExtJsServiceImpl.java (hunk 7 above)
Comment (axl8713) [reply src/modules/rest/extjs/src/main/java/it/geosolutions/geostore/services/rest/impl/RESTExtJsServiceImpl.java:353]: A dedicated `PermissionService` was created to handle resource access checks. However, the specific logic for setting permission flags has been intentionally kept separate within the respective services to ensure no existing functional requirements are inadvertently affected.

File: src/modules/rest/extjs/src/main/java/it/geosolutions/geostore/services/rest/impl/RESTExtJsServiceImpl.java (hunk 8 above)
Comment (axl8713) [reply src/modules/rest/extjs/src/main/java/it/geosolutions/geostore/services/rest/impl/RESTExtJsServiceImpl.java]: Made the `ResourceEnvelop` inner class to use the extracted `PermissionService`.

File: src/core/model/src/main/java/it/geosolutions/geostore/core/model/Resource.java (hunk 1 above)
Comment (axl8713) [reply src/core/model/src/main/java/it/geosolutions/geostore/core/model/Resource.java:125]: You're right.
I've created a DTO for the resource+permission data, aligning it to the `Ext` naming already in place.
Thanks!

File: src/modules/rest/extjs/src/main/java/it/geosolutions/geostore/services/rest/impl/RESTExtJsServiceImpl.java (hunk 6 above)
Comment (axl8713) [reply src/modules/rest/extjs/src/main/java/it/geosolutions/geostore/services/rest/impl/RESTExtJsServiceImpl.java]: Created a specific DTO sort object. Thanks.

File: src/core/services-api/src/main/java/it/geosolutions/geostore/services/ResourceService.java (hunk 2 above)
Comment (axl8713) [reply src/core/services-api/src/main/java/it/geosolutions/geostore/services/ResourceService.java]: Created a parameter object for the resource search.
Documented in the javadoc the usage of the search method.

**Review by afabiani — APPROVED** [review]
